presto-hive/src/main/java/com/facebook/presto/hive/BackgroundHiveSplitLoader.java

```diff
@@ -393,6 +401,25 @@ private static boolean shouldUseFileSplitsFromInputFormat(InputFormat<?, ?> inpu
                 .iterator();
     }
 
+    private List<InternalHiveSplit> getVirtuallyBucketedSplits(Path path, FileSystem fileSystem, InternalHiveSplitFactory splitFactory, int bucketCount)
+    {
+        // List all files in the partition and assign virtual bucket number to each of them
+        List<InternalHiveSplit> splitList = new ArrayList<>();
+        try {
+            new HiveFileIterator(path, fileSystem, directoryLister, namenodeStats, FAIL).forEachRemaining(file -> splitFactory.createInternalHiveSplit(file, getVirtualBucket(bucketCount, file.getPath()))
+                    .ifPresent(splitList::add));
+        }
+        catch (NestedDirectoryNotAllowedException e) {
+            // Fail here to be on the safe side. This seems to be the same as what Hive does
```

Comment:
This comment seems incorrect, since Hive doesn't have this virtual bucket behavior. This should probably be `NOT_SUPPORTED` or a new `HIVE_SUBDIRECTORY_NOT_SUPPORTED` with a message like
```java
"Hive table '%s' partition '%s' contains sub-subdirectory, which is not supported for virtual bucketing"
```

Reply:
Actually, I don't see why we need to fail here, since sub-directories will work fine with virtual bucketing. We should use `createInternalHiveSplitIterator()` which will have the normal behavior for directory walking.

Reply:
Yea, we should let the file iterator recursively walk down the partition directory with nested directory policy set to `RECURSE` instead of `FAIL`. Will fix it.